### PR TITLE
Add a 'local' pod mode to the pod-utils and mkpod to generate decorated pods suitable for local testing.

### LINKED ISSUE #13475: Better local testing of decorated ProwJobs using Phaino.
**What would you like to be added**:
Phaino currently constructs Docker commands to run a job locally based on the ProwJob spec instead of the decorated pod spec. This distinction results in decorated jobs behaving differently with Phaino than they do in CI. In particular:
- Git refs are not properly checked out
- $ARTIFACTS and $GOPATH are not provided.
- logs/metadata/artifacts are not exported.

I think the most complete way to address this and avoid skew in the future would be to add a "local" mode to the relevant pod utilities to make them copy to an output directory instead of uploading to GCS (or just skip upload and assume the directory where files are staged is an output directory). We can then update Phaino to use `mkpod` in addition to `mkpj` to get the decorated pod spec and then either:
1) Update Phaino to work with multiple containers
or
1) Update Phaino to use `kind` instead of docker.

This might be overkill though given that achieving the first two list items would otherwise only require some minor changes to Phaino and the third is not quite as important.

I do like the idea of having Phaino be a simple wrapper around `mkpj`, `mkpod`, and `kind` though. The requirements for that would be a local pod utility mode and the ability to mount volumes from the host into kind pods (this sounds easy?).

/assign
cc @fejta @stevekuznetsov @spiffxp

### PR DESCRIPTION
**Pull request by cjwagner**
This PR is broken into 3 commits:
- Allow mkpod to autogenerate buildID when 'snowflake' is specified.
- Add local output mode to pod utilities.
  - In this mode the pod utilities don't require or mount a GCS secret and don't upload to GCS. Instead a `hostPath` output volume is mounted and files that would have been uploaded to GCS are copied there.
- Add support for local pods to mkpod and include a script for running local pods with kind.
  - When mkpod is run with `--local` it configures the pod utilities to output locally and prompts for each volume that is not a `hostPath` or `emptyDir` to be replaced by an `emptyDir`, `hostPath`, or kept the same (this assumes the volume will be available in the cluster where the pod will run).
  - The included `local-kind.sh` script is a simple wrapper around `mkpj`, `mkpod`, and `kind`. It creates a kind cluster named `mkpod` if one doesn't exist and uses `mkpj` and `mkpod` to create a pod that is then applied to the kind cluster and watched.
    - I'm not set on this being a bash script, but I wrote it this way initially, because it makes it easy to modify for different workflows (e.g. any k8s resources the job depends on could be added to the script). I'm happy to change this.

Example usage after installing docker and kind:
```
./local-kind.sh pull-test-infra-verify-file-perms
```
Testing a job from a different Prow instance in a different repo (`istio/test-infra`):
```
CONFIG_PATH=$GOPATH/src/istio.io/test-infra/prow/config.yaml JOB_CONFIG_PATH=$GOPATH/src/istio.io/test-infra/prow/cluster/jobs ./local-kind.sh pull-test-infra-prow-checkconfig
```

I still need to add tests and documention, but I wanted to get some feedback on the direction first.
/assign @stevekuznetsov @fejta 
fixes #13475

### CODE REVIEW

File: prow/cmd/mkpod/local-kind-config.yaml [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# Example kind config for getting output on the machine running kind instead of the contain 'node'.
```
Comment (stevekuznetsov) [inline prow/cmd/mkpod/local-kind-config.yaml]: `contain 'node'`?

File: prow/cmd/mkpod/local-kind.sh [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+job=${1:-pull-test-infra-yamllint}
```
Comment (stevekuznetsov) [inline prow/cmd/mkpod/local-kind.sh]: Quote your var expansions

File: prow/cmd/mkpod/local-kind.sh [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+config=${CONFIG_PATH:-$(readlink -f $(dirname "${BASH_SOURCE[0]}")/../../config.yaml)}
+job_config=${JOB_CONFIG_PATH-$(readlink -f $(dirname "${BASH_SOURCE[0]}")/../../../config/jobs)}
+
+echo job=${job}
+echo config=${config}
+echo job_config=${job_config}
+
+if [[ -n ${job_config} ]]
```
Comment (stevekuznetsov) [inline prow/cmd/mkpod/local-kind.sh]: you enforce that this is set above with `:-` so no need to check

File: prow/cmd/mkpod/main.go [hunk 4]
```diff
@@ -82,13 +94,41 @@ func main() {
 		o.buildID = job.Status.BuildID
 	}
 
+	if strings.ToLower(o.buildID) == "snowflake" {
```
Comment (stevekuznetsov) [inline prow/cmd/mkpod/main.go]: Why not just let the user provide one and not use the special case here?

File: prow/gcsupload/options.go [hunk 5]
```diff
@@ -61,11 +61,18 @@ type Options struct {
 	// paths that are parsed to get more granular
 	// fields.
 	gcsPath gcs.Path
+
+	// LocalOutputDir specifies a directory where files should be copied INSTEAD of uploading to GCS.
+	// This option is useful for testing jobs that use the pod-utilities without actually uploading.
+	LocalOutputDir string `json:"local_output_dir,omitempty"`
```
Comment (stevekuznetsov) [inline prow/gcsupload/options.go]: I think I'd prefer us to implement an option on jobs to not do any GCS interaction instead of running a single-purpose tool (`gcsupload`) to do something explicitly not related to GCS or uploads :D

File: prow/cmd/mkpod/main.go (hunk 4 above)
Comment (cjwagner) [reply prow/cmd/mkpod/main.go]: We already let the user provide one if they want something specific, this doesn't change that. Most of the time the user doesn't care about the build ID though and just wants a unique one so forcing the user to provide a unique ID each time sounds unreasonable to me especially when we default to generating snowflake IDs in prod which is something mkpod can totally do.

File: prow/gcsupload/options.go (hunk 5 above)
Comment (cjwagner) [reply prow/gcsupload/options.go]: The reasoning for adding a local mode to the gcsupload utility instead of having a separate localoutput utility is to minimize skew between local testing and prod behavior. By reusing gcsupload we ensure we output the same files we would upload and perform all the same logic. Minimizing skew between local testing and prod is the motivation for this PR so I'm trying to prioritize that.
I don't really see this as not related to GCS uploads either, I see this more as a dry-run equivalent. It is supposed to simulate a GCS upload.

File: prow/cmd/mkpod/local-kind.sh (hunk 2 above)
Comment (cjwagner) [reply prow/cmd/mkpod/local-kind.sh]: Will do, I'm a bash noob and don't see how that would make a difference here though. Is the concern that without quotes `job` could be a list if there are spaces?

File: prow/cmd/mkpod/local-kind-config.yaml (hunk 1 above)
Comment (cjwagner) [reply prow/cmd/mkpod/local-kind-config.yaml]: `container 'node'` or `'node' container`. Spelling is hard.

**Comment by cjwagner** [comment]
Changed to having a general script and a prow.k8s.io specific one that defaults the config locations. We can create similar scripts for other prow instances.
PTAL.

File: config/pj-on-kind.sh [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+# Runs prow/pj-on-kind.sh with config arguments specific to the prow.k8s.io instance.
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+export CONFIG_PATH="$(readlink -f $(dirname "${BASH_SOURCE[0]}")/../prow/config.yaml)"
+export JOB_CONFIG_PATH="$(readlink -f $(dirname "${BASH_SOURCE[0]}")/jobs)"
+bash <(curl -s https://raw.githubusercontent.com/cjwagner/test-infra/local-pod-go/prow/pj-on-kind.sh) "$@"
```
Comment (fejta) [inline config/pj-on-kind.sh]: cjwagner?

Comment (cjwagner) [reply config/pj-on-kind.sh]: That was pointed at my fork for testing. Fixed.

File: config/pj-on-kind.sh [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+
+# Runs prow/pj-on-kind.sh with config arguments specific to the prow.k8s.io instance.
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+export CONFIG_PATH="$(readlink -f $(dirname "${BASH_SOURCE[0]}")/../prow/config.yaml)"
+export JOB_CONFIG_PATH="$(readlink -f $(dirname "${BASH_SOURCE[0]}")/jobs)"
+bash <(curl -s https://raw.githubusercontent.com/kubernetes/test-infra/master/prow/pj-on-kind.sh) "$@"
```
Comment (stevekuznetsov) [inline config/pj-on-kind.sh]: Why a `curl` instead of relative path?

File: prow/cmd/mkpod/local-kind-config.yaml (hunk 1 above)
Comment (stevekuznetsov) [reply prow/cmd/mkpod/local-kind-config.yaml]: outstanding?

File: prow/cmd/mkpod/main.go [hunk 8]
```diff
@@ -98,3 +138,47 @@ func main() {
...
+		prompt := `For each of the following volumes specify one of:
+ - 'empty' to use an emptyDir;
+ - a path on the host to use hostPath;
+ - '' (nothing) to use the existing volume source and assume it is available in the cluster`
+		fmt.Fprintln(os.Stderr, prompt)
+		for _, vol := range volsToFix {
+			fmt.Fprintf(os.Stderr, "Volume %q: ", vol.Name)
+
+			var choice string
+			fmt.Scanln(&choice)
```
Comment (stevekuznetsov) [inline prow/cmd/mkpod/main.go]: Often it's nice to allow this to be flag-driven so it can be scripted, if for instance the UX to deliver to devs is this tool wrapped with some more context (maybe create tempdirs for you etc)

File: prow/pj-on-kind.sh [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+# Prep and check args.
+job="${1:-""}"
+config="${CONFIG_PATH:-""}"
+job_config="${JOB_CONFIG_PATH:-""}"
+out_dir="${OUT_DIR:-/tmp/prowjob-out}"
```
Comment (stevekuznetsov) [inline prow/pj-on-kind.sh]: nit: if you're going to quote the empty strings for defaults above you might as well quote  this default, too :)

File: prow/pj-on-kind.sh [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+
+echo job=${job}
+echo CONFIG_PATH=${config}
+echo JOB_CONFIG_PATH=${job_config}
+echo OUT_DIR=${out_dir} "(May be different when reusing an existing kind cluster.)"
+
+if [[ -z ${job} ]]
```
Comment (stevekuznetsov) [inline prow/pj-on-kind.sh]: quote your expansions

File: prow/pj-on-kind.sh [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+set -o pipefail
+
+# Prep and check args.
+job="${1:-""}"
+config="${CONFIG_PATH:-""}"
+job_config="${JOB_CONFIG_PATH:-""}"
+out_dir="${OUT_DIR:-/tmp/prowjob-out}"
+
+echo job=${job}
+echo CONFIG_PATH=${config}
```
Comment (stevekuznetsov) [inline prow/pj-on-kind.sh]: quote strings with embedded expansions

File: prow/pj-on-kind.sh [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+  echo "Must specify config.yaml location via CONFIG_PATH env var."
+  exit 2
+fi
+if [[ -n ${job_config} ]]
+then
+  job_config="--job-config-path=${job_config}"
+fi
+
+# Install kind and set up cluster if not already done.
+if [[ -z $(which kind) ]]
```
Comment (stevekuznetsov) [inline prow/pj-on-kind.sh]: `which` will output something like `/usr/bin/which: no kind in (${PATH})` on failure so this does not do what you want.


use:

```sh
if ! which kind >/dev/null 2>&1; then
```

Comment (stevekuznetsov) [reply prow/pj-on-kind.sh]: and i think there was some weirdness with macos on this one, so you might have to use `command -v`

File: prow/pj-on-kind.sh [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+fi
+found="false"
+for clust in $(kind get clusters)
+do
+  if [[ ${clust} == "mkpod" ]]
+  then
+    found="true"
+  fi
+done
+if [[ ${found} == "false" ]]
```
Comment (stevekuznetsov) [inline prow/pj-on-kind.sh]: nit: could just inline this into the `if` where you find the match instead of doing it after the loop`

File: prow/pj-on-kind.sh [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+      - containerPath: /output
+        hostPath: ${out_dir}
+EOF
+
+  kind create cluster --name=mkpod --config=${PWD}/kind-config.yaml --wait=5m
+  rm ${PWD}/kind-config.yaml
+fi
+export KUBECONFIG="$(kind get kubeconfig-path --name="mkpod")"
+
+# Install mkpj and mkpod if not already done.
```
Comment (stevekuznetsov) [inline prow/pj-on-kind.sh]: maybe prefer running the published containers?

**Comment by stevekuznetsov** [comment]
GCSupload/ go code LGTM

File: config/pj-on-kind.sh (hunk 7 above)
Comment (cjwagner) [reply config/pj-on-kind.sh]: With `curl`, this script can be copied into any other repo with Prow config (e.g. `istio/test-infra`) and the only change needed to make everything work is to adjust the config envvars accordingly. Then, jobs for that Prow instance can be run with `./pj-on-kind.sh my-istio-job`. Fetching the `k/t-i/prow/pj-on-kind.sh` file each time ensures that the latest version is always used.

In this case a relative path is equivalent, but I want a copy-pasteable example.

Comment (cjwagner) [reply config/pj-on-kind.sh]: I changed to use a relative path, but left the commented out the `curl` version.

File: prow/cmd/mkpod/local-kind.sh (hunk 2 above)
Comment (stevekuznetsov) [reply prow/cmd/mkpod/local-kind.sh]: Yes. Always quote them so that the absence of quotes is an intentional choice and can be interpreted as such.

File: prow/pj-on-kind.sh [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+  fi
+  if [[ -n "${job_config}" ]]; then
+    job_config="--job-config-path=${job_config}"
+  fi
+}
+
+# Ensures installation of prow tools, kind, and a kind cluster named "mkpod".
+function ensureInstall() {
+  # Install mkpj and mkpod if not already done.
+  if [[ $(command -v mkpj &>/dev/null) ]]; then
```
Comment (stevekuznetsov) [inline prow/pj-on-kind.sh]: If you're checking the result of a call just do

```sh
if ! command -v mkpj >/dev/null 2>&1; then
```

No need to test if the output string is or is not empty

File: config/pj-on-kind.sh (hunk 7 above)
Comment (stevekuznetsov) [reply config/pj-on-kind.sh]: Might want to just package it in a container image for re-use
